File: ios/RNMediaPlayer/RNMediaPlayer.m
//
//  RNMediaPlayer.m
//  RNMediaPlayer
//
//  Created by sean on 2022/3/10.
//  Copyright © 2022 ZJ. All rights reserved.
//

#import "RNMediaPlayer.h"
#import <ZJMediaPlayer/ZJMediaPlayer.h>
#import "RNMonitorManager.h"

NSString *const kOnMediaPlayerPlayStatus = @"onMediaPlayerPlayStatus";
NSString *const kOnMediaPlayerStatisticalInfo = @"onMediaPlayerStatisticalInfo";
NSString *const kOnMediaPlayerRecordStatus = @"onMediaPlayerRecordStatus";

@interface RNMediaPlayer () <ZJMediaPlayerDelegate>

@property (nonatomic, assign) BOOL isHasListeners;
@property (nonatomic, strong) ZJMediaPlayer *mediaPlayer;

@end

@implementation RNMediaPlayer
RCT_EXPORT_MODULE(MediaPlayer);

- (void)dealloc {
    if (_mediaPlayer) {
        _mediaPlayer.delegate = nil;
        _mediaPlayer.glMonitor = nil;
        _mediaPlayer = nil;
    }
}

- (void)startObserving {
    self.isHasListeners = YES;
}

- (void)stopObserving {
    self.isHasListeners = NO;
}

- (NSMutableDictionary *)emptyBody {
    NSMutableDictionary *body = @{}.mutableCopy;
    return body;
}

- (void)sendEvent:(NSString *)name body:(NSMutableDictionary *)body {
    if (!self.isHasListeners) return;
    [self sendEventWithName:name body:body];
}

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[kOnMediaPlayerPlayStatus, kOnMediaPlayerStatisticalInfo, kOnMediaPlayerRecordStatus];
}

- (NSDictionary *)constantsToExport {
    return @{kOnMediaPlayerPlayStatus: kOnMediaPlayerPlayStatus,
             kOnMediaPlayerStatisticalInfo: kOnMediaPlayerStatisticalInfo,
             kOnMediaPlayerRecordStatus: kOnMediaPlayerRecordStatus
    };
}

- (ZJMediaPlayer *)mediaPlayer {
    if (!_mediaPlayer) {
        _mediaPlayer = [[ZJMediaPlayer alloc] init];
    }
    return _mediaPlayer;
}

#pragma mark -

RCT_EXPORT_METHOD(openDebugMode) {
    [[ZJAuthManager shared] openDebugMode];
}

RCT_EXPORT_METHOD(configWithKey:(NSString *)key secret:(NSString *)secret cb:(RCTResponseSenderBlock)cb) {
    BOOL ret = [[ZJAuthManager shared] configWithKey:key secret:secret];
    cb(@[@(ret)]);
}

RCT_EXPORT_METHOD(isAuthOK:(RCTResponseSenderBlock)cb) {
    cb(@[@([[ZJAuthManager shared] isAuthOK])]);
}

#pragma mark -

RCT_EXPORT_METHOD(playerId:(RCTResponseSenderBlock)cb) {
    cb(@[@(self.mediaPlayer.hash)]);
}

RCT_EXPORT_METHOD(addMonitor:(NSInteger)viewId) {
    RNZJGLMonitor *view = [[RNMonitorManager shared] getMonitor:viewId];
    if (!view) return;
    self.mediaPlayer.glMonitor = view;
    view.mediaPlayer = self.mediaPlayer;
}

RCT_EXPORT_METHOD(start:(NSString *)url) {
    self.mediaPlayer.delegate = self;
    [self.mediaPlayer start:url];
}

RCT_EXPORT_METHOD(restart) {
    [self.mediaPlayer restart];
}

RCT_EXPORT_METHOD(seek:(NSInteger)msTime) {
    [self.mediaPlayer seek:msTime];
}

RCT_EXPORT_METHOD(stop) {
    self.mediaPlayer.delegate = nil;
    [self.mediaPlayer stop];
}

RCT_EXPORT_METHOD(openVideo) {
    [self.mediaPlayer openVideo];
}

RCT_EXPORT_METHOD(openAudio) {
    [self.mediaPlayer openAudio];
}

RCT_EXPORT_METHOD(closeVideo) {
    [self.mediaPlayer closeVideo];
}

RCT_EXPORT_METHOD(closeAudio) {
    [self.mediaPlayer closeAudio];
}

RCT_EXPORT_METHOD(mute:(BOOL)mute) {
    self.mediaPlayer.mute = mute;
}

RCT_EXPORT_METHOD(setMediaCacheTime:(NSInteger)cacheTime pursueTime:(NSInteger)pursueTime) {
    [self.mediaPlayer setMediaCacheTime:cacheTime pursueTime:pursueTime];
}

RCT_EXPORT_METHOD(setDenoiseLevel:(NSInteger)nLevel) {
    [self.mediaPlayer setDenoiseLevel:nLevel];
}

RCT_EXPORT_METHOD(setMediaSyncMode:(BOOL)sync) {
    [self.mediaPlayer setMediaSyncMode:sync];
}

RCT_EXPORT_METHOD(startRecord:(NSString *)filePath) {
    [self.mediaPlayer startRecord:filePath];
}

RCT_EXPORT_METHOD(stopRecord) {
    [self.mediaPlayer stopRecord];
}

RCT_EXPORT_METHOD(isRecording:(RCTResponseSenderBlock)cb) {
    cb(@[@([self.mediaPlayer isRecording])]);
}

RCT_EXPORT_METHOD(getRecordDuration:(RCTResponseSenderBlock)cb) {
    cb(@[@([self.mediaPlayer getRecordDuration])]);
}

#pragma mark - ZJMediaPlayerDelegate

- (void)didMediaPlayerPlay:(ZJMediaStreamPlayer *_Nonnull)player status:(ZJ_MEDIA_PLAY_STATUS)status errCode:(NSInteger)errCode {
    NSMutableDictionary *dic = [self emptyBody];
    dic[@"playerId"] = @(player.hash);
    dic[@"status"] = @(status);
    dic[@"errCode"] = @(errCode);
    [self sendEvent:kOnMediaPlayerPlayStatus body:dic];
}

- (void)didMediaPlayerStatisticalInfo:(ZJMediaStreamPlayer *_Nonnull)player statInfo:(ZJMediaStatisticalInfo *)statInfo {
    NSMutableDictionary *dic = [self emptyBody];
    dic[@"playerId"] = @(player.hash);
    dic[@"videoWidth"] = @(statInfo.videoWidth);
    dic[@"videoHeight"] = @(statInfo.videoHeight);
    dic[@"videoFPS"] = @(statInfo.videoFPS);
    dic[@"videoBps"] = @(statInfo.videoBps);
    dic[@"audioBps"] = @(statInfo.audioBps);
    dic[@"timestamp"] = @(statInfo.timestamp);
    dic[@"videoTotalFrame"] = @(statInfo.videoTotalFrame);
    dic[@"videoDropFrame"] = @(statInfo.videoDropFrame);
    dic[@"onlineCount"] = @(statInfo.onlineCount);
    [self sendEvent:kOnMediaPlayerStatisticalInfo body:dic];
}

- (void)didMediaPlayerRecord:(ZJMediaStreamPlayer *_Nonnull)player status:(ZJ_MEDIA_RECORD_STATUS)status filePath:(NSString *_Nullable)filePath errCode:(NSInteger)errCode {
    NSMutableDictionary *dic = [self emptyBody];
    dic[@"playerId"] = @(player.hash);
    dic[@"status"] = @(status);
    dic[@"errCode"] = @(errCode);
    dic[@"filePath"] = filePath;
    [self sendEvent:kOnMediaPlayerRecordStatus body:dic];
}

@end

File: android/src/main/java/com/eafy/rn/RNMediaPlayer.java
package com.eafy.rn;

import androidx.annotation.Nullable;

import com.eafy.zjmediaplayer.Listener.ZJDisplayRatioType;
import com.eafy.zjmediaplayer.Listener.ZJMediaPlayerListener;
import com.eafy.zjmediaplayer.ZJAuthManager;
import com.eafy.zjmediaplayer.ZJMediaPlayer;
import com.eafy.zjmediaplayer.ZJMediaStatisticalInfo;
import com.eafy.zjmediaplayer.ZJMediaStreamPlayer;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.uimanager.events.RCTEventEmitter;

import java.util.HashMap;
import java.util.Map;

public class RNMediaPlayer extends ReactContextBaseJavaModule implements ZJMediaPlayerListener {

    public static final String kOnMediaPlayerPlayStatus = "onMediaPlayerPlayStatus";
    public static final String kOnMediaPlayerStatisticalInfo = "onMediaPlayerStatisticalInfo";
    public static final String kOnMediaPlayerRecordStatus = "onMediaPlayerRecordStatus";

    private ReactApplicationContext reactContext = null;
    private ZJMediaPlayer mediaPlayer = null;
    public RNZJGLMonitor monitor = null;

    public RNMediaPlayer(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "MediaPlayer";
    }

    @Nullable
    @Override
    public Map<String, Object> getConstants() {
        Map<String, Object> constants = new HashMap<>();
        constants.put(kOnMediaPlayerPlayStatus, kOnMediaPlayerPlayStatus);
        constants.put(kOnMediaPlayerStatisticalInfo, kOnMediaPlayerStatisticalInfo);
        constants.put(kOnMediaPlayerRecordStatus, kOnMediaPlayerRecordStatus);
        return constants;
    }

    private void pushEvent(String name, WritableMap data) {
        if (reactContext == null) return;
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(name, data);
    }

    public ZJMediaPlayer mediaPlayer() {
        if (mediaPlayer != null) return mediaPlayer;

        mediaPlayer = new ZJMediaPlayer(reactContext.getApplicationContext());
        return mediaPlayer;
    }

    @ReactMethod
    public void openDebugMode() {
        ZJAuthManager.shared().openDebugMode();
    }

    @ReactMethod
    public void configWithKey(String key, String secret, Promise cb) {
        boolean ret = ZJAuthManager.shared().config(reactContext.getApplicationContext(), key, secret);
        if (cb != null) {
            cb.resolve(ret);
        }
    }

    @ReactMethod
    public void isAuthOK(Promise cb) {
        if (cb == null) return;
        cb.resolve(ZJAuthManager.shared().isDebugMode());
    }

    @ReactMethod
    public void playerId(Promise cb) {
        if (cb == null) return;
        cb.resolve(mediaPlayer().hashCode());
    }

    @ReactMethod
    public void addMonitor(int viewId) {
        RNZJGLMonitor monitor = RNMonitorManager.shared().getMonitor(viewId);
        if (monitor == null) return;
        mediaPlayer().addMonitor(monitor);
        if (this.monitor != null) {
            this.monitor.mediaPlayer = null;
        }
        this.monitor = monitor;
        monitor.mediaPlayer = this;
    }

    @ReactMethod
    public void start(String url) {
        mediaPlayer().setListener(this);
        this.mediaPlayer.startURL(url);
    }

    @ReactMethod
    public void restart() {
        mediaPlayer().restart();
    }

    @ReactMethod
    public void seek(int msTime) {
        mediaPlayer().seek(msTime);
    }

    @ReactMethod
    public void stop() {
        mediaPlayer().setListener(null);
        mediaPlayer().stop();
    }

    @ReactMethod
    public void openVideo() {
        mediaPlayer().openVideo();
    }

    @ReactMethod
    public void closeVideo() {
        mediaPlayer().closeVideo();
    }

    @ReactMethod
    public void mute(boolean mute) {
        mediaPlayer().setMute(mute);
    }

    @ReactMethod
    public void setMediaCacheTime(int cacheTime, int pursueTime) {
        mediaPlayer().setMediaCacheTime(cacheTime, pursueTime);
    }

    @ReactMethod
    public void setDenoiseLevel(int nLevel, float gainVolume) {
        mediaPlayer().setDenoiseLevel(nLevel, gainVolume);
    }

    @ReactMethod
    public void setMediaSyncMode(boolean sync) {
        mediaPlayer().setMediaSyncMode(sync);
    }

    @ReactMethod
    public void startRecord(String path) {
        mediaPlayer().startRecord(path);
    }

    @ReactMethod
    public void stopRecord() {
        mediaPlayer().stopRecord();
    }

    @ReactMethod
    public void isRecording(Promise cb) {
        if (cb == null) return;
        cb.resolve(mediaPlayer().isRecording());
    }

    @ReactMethod
    public void getRecordDuration(Promise cb) {
        if (cb == null) return;
        cb.resolve(mediaPlayer().getRecordDuration());
    }

    /*************************   ZJMediaPlayerListener   *****************************/

    @Override
    public void didMediaPlayerPlay(ZJMediaStreamPlayer player, int status, int errCode) {
        WritableMap map = Arguments.createMap();
        map.putInt("playerId", player.hashCode());
        map.putInt("status", status);
        map.putInt("errCode", errCode);
        pushEvent(kOnMediaPlayerPlayStatus, map);
    }

    @Override
    public void didMediaPlayerStatisticalInfo(ZJMediaStreamPlayer player, ZJMediaStatisticalInfo statInfo) {
        WritableMap map = Arguments.createMap();
        map.putInt("playerId", player.hashCode());
        map.putInt("videoWidth", statInfo.videoWidth);
        map.putInt("videoHeight", statInfo.videoHeight);
        map.putInt("videoFPS", statInfo.videoFPS);
        map.putInt("videoBps", statInfo.videoBps);
        map.putInt("audioBps", statInfo.audioBps);
        map.putInt("timestamp", (int)statInfo.timestamp);
        map.putInt("videoTotalFrame", statInfo.videoTotalFrame);
        map.putInt("videoDropFrame", statInfo.videoDropFrame);
        map.putInt("onlineCount", statInfo.onlineCount);
        pushEvent(kOnMediaPlayerStatisticalInfo, map);
    }

    @Override
    public void didMediaPlayerRecord(ZJMediaStreamPlayer player, int status, String filePath, int errCode) {
        WritableMap map = Arguments.createMap();
        map.putInt("playerId", player.hashCode());
        map.putInt("status", status);
        map.putInt("errCode", errCode);
        map.putString("filePath", filePath);
        pushEvent(kOnMediaPlayerRecordStatus, map);
    }
}
